models/authentication/agent/organisation/index.ts

```diff
@@ -0,0 +1,88 @@
+import { HttpForbiddenError } from '#clients/exceptions';
+import {
+  CanRequestAuthorizationException,
+  NoSiretException,
+} from '#models/authentication/authentication-exceptions';
+import { isServicePublic, IUniteLegale } from '#models/core/types';
+import { getUniteLegaleFromSlug } from '#models/core/unite-legale';
+import { extractSirenFromSiret, Siren } from '#utils/helpers';
+import { defaultAgentScopes } from '../scopes/default-agent-scopes';
+import { mightBeAnAdministration } from './might-be-an-administration';
+
+const basicOrganisationHabilitation = {
+  scopes: [...defaultAgentScopes],
+  userType: 'Agent connecté',
+  isSuperAgent: false,
+};
+
+export class AgentOrganisation {
+  private isFromMCP: boolean;
+  private siren: Siren;
+
+  constructor(private domain: string, private idpId: string, siret: string) {
+    this.isFromMCP = this.isMCP(idpId);
+    this.siren = extractSirenFromSiret(siret);
+  }
+
+  async getHabilitationLevel() {
+    if (!this.siren) {
+      throw new NoSiretException(
+        'The user doesn‘t have a siret',
+        `${this.domain} - ${this.idpId} - ${
+          this.isFromMCP ? 'ProConnectIdentité' : 'FI ministères'
+        }`
+      );
+    }
+
+    const uniteLegale = await getUniteLegaleFromSlug(this.siren, {
+      page: 0,
+      isBot: false,
+    });
+
+    const isAnAdministration = isServicePublic(uniteLegale);
+    if (isAnAdministration) {
+      if (!this.isAdministrationAuthorized(uniteLegale)) {
+        throw new CanRequestAuthorizationException(
+          uniteLegale.natureJuridique,
+          this.siren
+        );
+      } else if (this.isAdministrationTrustworthy()) {
+        // return trustworthytOrganisationHabilitation;
+        return basicOrganisationHabilitation;
+      }
+      return basicOrganisationHabilitation;
+    }
+
+    if (mightBeAnAdministration(uniteLegale.natureJuridique)) {
+      throw new CanRequestAuthorizationException(
+        uniteLegale.natureJuridique,
+        this.siren
+      );
+    }
+
+    throw new HttpForbiddenError('Organization is not a service public');
+  }
```

Comment:
Question :
If you start with this condition `if (mightBeAnAdministration(uniteLegale.natureJuridique))`, you wouldn't need to define a `isAdministrationAuthorized` because the list in question is always to be included in `might be authorised` anyways(talking about ['4110', '4120', '4140', '4150', '7381', '7410']), no?

Reply:
Your implementation suggestion is much more efficient, I like it

Reply:
Actually it does not work well as some siren might belong to the whitelist at siren level and not at nature juridique

Testing first mightBeAnAdministration will block those